src/interop/js-interop/js-types.md

```diff
@@ -0,0 +1,242 @@
+---
+title: JS types
+description: Usage information about the core types in JS interop.
+---
+
+Dart values and JS values belong to separate language domains. When compiling to
+Wasm, they execute in separate *runtimes* as well. As such, you should treat JS
+values as foreign types. To provide Dart types for JS values,
+[`dart:js_interop`] exposes a set of [extension types] prefixed with `JS` called
+"JS types".
+
+Importantly, these types have compiler-specific [representation types]. This
+means that their runtime type will differ based on whether code is compiled to
+Wasm or JS. In order to interact with and examine these JS values, you should
+use [`external`] interop members.
+
+## Type hierarchy
+
+JS types form a natural type hierarchy:
+
+- Top type: `JSAny`, which is any non-nullish JS value
+  - Primitives: `JSNumber`, `JSBoolean`, `JSString`
+  - `JSSymbol`
+  - `JSBigInt`
+  - `JSObject`, which is any JS object
+    - `JSFunction`
+      - `JSExportedDartFunction`, which represents a Dart callback that was
+      converted to a JS function
+    - `JSArray`
+    - `JSPromise`
+    - `JSDataView`
+    - `JSTypedArray`
+      - JS typed arrays like `JSUint8Array`
+    - `JSBoxedDartObject`, which allows users to box and pass Dart values
+      opaquely within the same Dart runtime
+
+You can find the definition of each type in the [`dart:js_interop` API docs].
+
+{% comment %}
+TODO (srujzs): Should we add a tree diagram instead for JS types?
+{% endcomment %}
+
+## Conversions
+
+To use a value from one domain to another, you will likely want to *convert* the
+value to the corresponding type of the other domain. For example, you might want
+to convert a Dart `List<JSString>` into a JS array of strings, which is
+represented by the JS type `JSArray<JSString>`, so that you can pass the array
+to a JS interop API.
+
+Dart supplies a number of `extension` members on various Dart types and JS types
+to convert the values between the domains for you.
+
+Members that convert values from Dart to JS are usually labeled-as or
+prefixed-with `toJS`:
+
+```dart
+String str = 'hello world';
+JSString jsStr = str.toJS;
+```
+
+Members that convert values from JS to Dart are usually labeled-as or
+prefixed-with `toDart`:
+
+```dart
+JSNumber jsNum = ...;
+int integer = jsNum.toDartInt;
+```
+
+These conversion functions are often marked `external`, but aren't interop
```

Comment:
There's an [open PR](https://github.com/dart-lang/site-www/pull/5481) right now for defining `external` in the docs (I'll merge that soon). I can add these details to that, and then can we just link to that section here? That could help avoid going into too much detail here.

I support dropping it just because the example code right above doesn't show it with `external`, so that was just confusing for me to read. If it's the more common scenario, maybe put `external` in the example code?

Reply:
My general concern was that users would look at the API surface in js_interop.dart, see that there are interop members on JS types that are marked `external`, but then also see that other members are marked as `external` that don't follow the same semantics. Maybe the better way to solve this is to add/require the `@patch` annotation on the user-facing declaration so it's obvious that it's different from other `external` members.

I'm okay with removing this as it doesn't really dive into that difference anyways, so only half-explaining it is possibly worse than not mentioning it.

> doesn't show it with external, so that was just confusing for me to read. If it's the more common scenario, maybe put external in the example code?

The conversion functions being used are `external` e.g. `String.toJS`, but if we're highlighting their usage, the example wouldn't show the `external` as that's tied to the declaration. Maybe linking to the declaration would reduce the confusion, but this sentence is being removed, so nevermind. :)